README.md
# AWS Serverless VideoConverter

![diagram](docs/diagram.png)

## About

AWS Serverless VideoConverter is an open to use solution that can convert video in S3 with MediaConvert. The converter job can be executed manually by queue from SQS or S3 event notification. All tasks are stored in DynamoDB and status will be auto updated with MediaConvert CloudWatch event.



This project contains source code and supporting files that can deploy with the SAM CLI. It includes the following files and folders.

- video_transfer - Source codes for the application's Lambda function.
- docs - Addon documents, such as diagram, media info output structure, etc..
- events -Example events used for local debug.
- layers -Lambda layers of application.
- samples - Sample file can used for CLI to config your runtime environment.
- template.yaml - A template that defines the application's AWS resources.



The application uses several AWS resources, including Lambda functions, DynamoDB tables and SQS by default.
These resources are defined in the `template.yaml` file in this project, your can as your wish.

## Develop

This application’s codes are using Python3, you can use any favor tool to code with them.

If you prefer to use an integrated development environment (IDE) to build and test your application, you can use the AWS Toolkit. See the following links to get started.

* [CLion](https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/welcome.html)
* [GoLand](https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/welcome.html)
* [IntelliJ](https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/welcome.html)
* [WebStorm](https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/welcome.html)
* [Rider](https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/welcome.html)
* [PhpStorm](https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/welcome.html)
* [PyCharm](https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/welcome.html)
* [RubyMine](https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/welcome.html)
* [DataGrip](https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/welcome.html)
* [VS Code](https://docs.aws.amazon.com/toolkit-for-vscode/latest/userguide/welcome.html)
* [Visual Studio](https://docs.aws.amazon.com/toolkit-for-visual-studio/latest/user-guide/welcome.html)



The main resources and codes are list bellow, more details can check the comments in sources.

```shell
.
├── template.yml  # The SAM template files, includes all required AWS resources defined.
├── video_converter
│   └── __init__.py
│   ├── auto_executor.py   # The lambda function with S3 notification and start a converter job
│   └── manual_executor.py # The lambda function with SQS message and start converter job(s)
│   └── mediainfo.py       # The helper classes for mediainfo 
│   └── requirement.txt    # The python pip install requirements
│   └── task.py            # The core function to handle task and converter job
│   └── task_event.py      # The lambda function to handle MediaConvert event to update task status
│   └── task_params.py     # The json params used to create a MediaConvert job
```

## Deploy

To deploy this application, you should use the Serverless Application Model Command Line Interface (SAM CLI), SAM CLI is an extension of the AWS CLI that adds functionality for building and testing Lambda applications. 

To use the SAM CLI, you need the following tools.

* AWS CLI - [How to Install](https://docs.aws.amazon.com/zh_cn/cli/latest/userguide/cli-chap-install.html)
* SAM CLI - [How to Install](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html)
* Python3.8 (Only for development) - [Download Python](https://www.python.org/downloads/)

> To deploy this application you must have a AKSK with IAM role has full access to Lambda, DynamoDB and SQS.

Before use the SAM CLI, you should set up you aws cli credential first, run the following in your shell:

```bash
aws configure
```

For more information, please see the [AWS CLI document](https://docs.aws.amazon.com/zh_cn/cli/latest/userguide/cli-chap-configure.html).

To deploy your application for the first time, run the following in your shell:

```bash
sam build
sam deploy --guided --capabilities CAPABILITY_IAM CAPABILITY_NAMED_IAM
```

The first command will build the source of your application. The second command will package and deploy your application to AWS, with a series of prompts:

* **Stack Name**

  The name of the stack to deploy to CloudFormation. This must be unique to your account and region.

* **AWS Region**

  The AWS region you want to deploy to.

* **Parameter DefaultJobTemplateName**

  The MediaConvert job template name used as default job template. This must in the same account and region, it can be created by your self anytime. 

* **Confirm changes before deploy**

* If set to yes, any change sets will be shown to you before execution for manual review. If set to no, the AWS SAM CLI will automatically deploy application changes.

* **Allow SAM CLI IAM role creation**

  Many AWS SAM templates, including this example, create AWS IAM roles required for the AWS Lambda function(s) included to access AWS services. By default, these are scoped down to minimum required permissions.

* **Save arguments to samconfig.toml**

  If set to yes, your choices will be saved to a configuration file inside the project, so that in the future you can just re-run `sam deploy` without parameters to deploy changes to your application.

## Config

After success deployed, you can config the application by set/update the items in DynamoDB table `video-converter-options`.

The options in table must have two fields, `Key` and `Value`. 

Currently, you can set following options use special keys:

* **MediaConvertRole**

* The role created by the application. This is used when create a MediaConvert job.

* **default-JobTemplateName**

  The default MediaConvert job template name used to create converter job. This is used when you not set a converter job template from your task define.

* **default-OutputBucket**

  The default bucket used for MediaConvert job output. This is used when you not set an output for your task.

  To use this option, the `bucket` must exists.
  
* **`bucket`-JobTemplateName**

  The bucket MediaConvert job template name used to create converter job. This is used when the job is auto executed from S3 notification.

  `bucket` must be the same as your S3 bucket name with case-sensitive.
  
  If this option not exists, the `default-JobTemplateName` will used as default.
  
* **`bucket`-OutputBucket**

  The bucket used for MediaConvert job output for a special bucket. This is used when the job is auto executed from S3 notification.

  `bucket` must be the same as your S3 bucket name with case-sensitive.

  If this option not exists, the `default-OutputBucket` will used as default.

## Run

To make the job auto executed when a new video file put in your S3 bucket, you can simply set a S3 event notification on your bucket. You can do it in your AWS console or use the AWS CLI shell:

```bash
aws s3api put-bucket-notification-configuration --bucket your-bucket-name --notification-configuration file://samples/bucket-notification-config.json
```

> * Please replace the **`THE-ARN-OF-AutomationFunction`** in the config file use the Arn value of the AutomationFunction. You can get the value from the deployment cli outputs or CloudFormation console.
> * Please make sure replace the **`your-bucket-name`** in command to the bucket name you want to auto execute the convert job.
> * **ATTENTION**: DO NOT use the source bucket as MediaConvert output target, this will cause infinite loop execution and increase your account cost.
>

For more details to use the configuration file, please see the [AWS CLI Document](https://docs.aws.amazon.com/cli/latest/reference/s3api/put-bucket-notification-configuration.html).



To start a manual job, just send a message to the `VideoConverterSQS`. You can do it in your AWS console or use the AWS CLI shell:

```shell
aws sqs send-message --queue-url the-videoconvertersqs-url --message-body "create task" --message-attributes file://samples/manual-task.json
```

> * Please make sure replace the **`the-videoconvertersqs-url`** in command to your SQS url, you can get it from the stack outputs as `VideoConverterSQS`.
>
> * Please make sure the config json is using the correct options such as bucket name, s3 object key and filter etc.

The attributes in manual-task.json are includes:

* **Bucket**

  The name of bucket which the task’s source video file(s) in. This attribute is *REQUIRED*.

* **Key**

  The s3 object key of the source video. If the key string ends with '`/`', will be using as a directory to look up objects in.

  This attribute is optional, if you don’t use it, just remove the whole entry.

* **Filter**

  The `JMESPath` expression to filter files in the bucket. 

  For example, to filter files with date, the expression can be:

  ​	`Contents[?to_string(LastModified)>='\"2021-01-01 00:00:00+00:00\"'][]`

  This attribute is optional, if you don’t use it, just remove the whole entry.

* **Force**

  Force creating task even it is already exists. Default is `false`.

  This attribute value can be `true` or `false`.

>  **ATTENTION**
>
> The lambda functions have default timeout setting, please make sure you never reach the limit. You can check the `template.yaml` to modify them base your requirement.

## Debug

The application can debug locally with the `sam ` command. To debug you should build it with the `sam build` command first.

```bash
sam build
```

The SAM CLI installs dependencies defined in `video_converter/requirements.txt`, creates a deployment package, and saves it in the `.aws-sam/build` folder.

Test a single function by invoking it directly with a test event. An event is a JSON document that represents the input that the function receives from the event source. Test events are included in the `events` folder in this project.

Run functions locally and invoke them with the `sam local invoke` command.

```bash
sam local invoke ManualFunction --event events/manualfunction.json
```

## Using MediaInfo

The application contains a layer which includes [MediaInfo](https://mediaarea.net/en/MediaInfo) runtime, it can be used to detect the media information in lambda. For example:

```python
from mediainfo import get_media_info

mi = get_media_info(bucket, key)
for track in mi.tracks:
    if track.track_type == 'Video':
        logger.info('Video resolution is %d x %d' % (track.width, track.high))
```

All lambda functions (exclude InitFunction) are referred this layer by default.

## Cleanup

To delete the application that you created, use the AWS CLI. Assuming you used your project name for the stack name, you can run the following:

```bash
aws cloudformation delete-stack --stack-name <the-stack-name-you-deployed>
```

> **NOTES**: This command just delete the application stack only, all converted jobs and output files are keep remained! 

## Resources

Reference documents to use and custom this application can be found at:

* [Boto3 Documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/index.html)
* [AWS Command Line Interface Documentation](https://docs.aws.amazon.com/cli/)
* [AWS Serverless Application Model Developer Guide](https://docs.aws.amazon.com/serverless-application-model/index.html)
* [AWS CloudFormation User Guide](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/Welcome.html)
* [AWS Lambda Documentation](https://docs.aws.amazon.com/lambda/)
* [AWS MediaConvert Documentation](https://docs.aws.amazon.com/mediaconvert/)
* [Amazon Simple Storage Service Documentation](https://docs.aws.amazon.com/s3/)
* [Amazon Simple Queue Service Documentation](https://docs.aws.amazon.com/sqs/)
* [Amazon DynamoDB Documentation](https://docs.aws.amazon.com/dynamodb/)
* [JMEPath Documentation](https://jmespath.org/)



CHANGELOG.md

<a name="v1.0.2"></a>
## [v1.0.2](https://github.com/SolardiaX/AWS-Serverless-VideoConverter/compare/v1.0.1...v1.0.2) (2021-10-14)

### Chore

* disable urllib3 InsecureRequestWarning
* fix typo

### Docs

* fix JMESPath example issue in README
* add CONTRIBUTING
* add CHANGELOG

### Feat

* add MediaInfo support
* enable dir as Key parameter for ManualFunction

### Test

* add input examples for local debug


<a name="v1.0.1"></a>
## [v1.0.1](https://github.com/SolardiaX/AWS-Serverless-VideoConverter/compare/v1.0.0...v1.0.1) (2021-10-12)

### Chore

* update gitignore

### Docs

* reference for JMEPath
* template for git-chglog

### Feat

* enhance MediaConvert event handler
* create LogGroups with SAM
* add SQS timeout property

### Fix

* issues with datetime & progress
* Force parameter not affect


<a name="v1.0.0"></a>
## v1.0.0 (2021-10-05)

### Docs

* update readme
* update api documents
* update readme

### Feat

* initial commit



video_converter/task.py
# -*- coding: utf-8 -*-

import boto3
import json
import uuid
from datetime import datetime

import urllib3
urllib3.disable_warnings()  # disable InsecureRequestWarning

task_table_name = 'video-converter-tasks'
taskitem_table_name = 'video-converter-task-items'
options_table_name = 'video-converter-options'

mc_client = boto3.client('mediaconvert')
mc_endpoints = mc_client.describe_endpoints()
converter = boto3.client('mediaconvert', endpoint_url=mc_endpoints['Endpoints'][0]['Url'], verify=False)
db = boto3.resource('dynamodb')


class Task:
    """
    Task of VideoConverter, refer to the table `'video-converter-tasks'` in DynamoDB
    """
    def __init__(self):
        self.taskId = None
        """ Task Id """
        self.bucket = None
        """ Bucket name of the source """
        self.template_name = None
        """ Template name of task, refer to the name of `JobTemplate` in MediaConvert """
        self.key = None
        """ 
        Key name of the source in bucket.
        If the attribute is None, will look up all objects in the bucket to create jobs
        Default is `None`
        """
        self.filter = None
        """ 
        Condition to filter objects in bucket
        This attribute MUST be a JMESPath style condition
        Default is `None`
        """
        self.executedAt = None
        """ Datetime in string that the task start to execute """
        self.total = 0
        """ Total jobs related to this task """
        self.finished = 0
        """ Total finished jobs related to this task """
        self.running = 0
        """ Total running jobs related to this task """
        self.error = 0
        """ Total error jobs related to this task """

    def as_dict(self) -> dict:
        """ A dict of task """
        return {
            'S_TaskId': self.taskId,
            'S_Bucket': self.bucket,
            'S_TemplateName': self.template_name,
            'S_Key': self.key,
            'S_Filter': self.filter,
            'S_ExecutedAt': self.executedAt,
            'N_Total': self.total,
            'N_Finished': self.finished,
            'N_Running': self.running,
            'N_Error': self.error,
        }

    @classmethod
    def from_item(cls, item: dict):
        """ Create a task instance from dict """
        task = cls()

        task.taskId = item['S_TaskId']
        task.bucket = item['S_Bucket']
        task.template_name = item.get('S_TemplateName', None)
        task.key = item.get('S_Key', None)
        task.filter = item.get('S_Filter', None)
        task.executedAt = item.get('S_ExecutedAt', None)
        task.total = item.get('N_Total', 0)
        task.finished = item.get('N_Finished', 0)
        task.running = item.get('N_Running', 0)
        task.error = item.get('N_Error', 0)

        return task


class TaskItem:
    """
    Task item of VideoConverter, refer to the table `'video-converter-task-items'` in DynamoDB
    It's also include infomation of the job in MediaConvert
    """
    def __init__(self):
        self.itemid = None
        """ Item Id """
        self.source = None
        """ Source of task item. For example: s3://bucket/sub-dir/example.mp4 """
        self.taskid = None
        """ Task Id of the item """
        self.status = None
        """ Status of the item, refer to the job status of MediaConvert """
        self.progress = 0
        """ Progress of the item, refer to the job progress of MediaConvert """
        self.created_at = None
        """ Datetime in string the item job created at """
        self.finished_at = None
        """ Datetime in string the item job finished at """
        self.error = None

    def as_dict(self):
        """ A dict of task item """
        return {
            'S_ItemId': self.itemid,
            'S_Source': self.source,
            'S_TaskId': self.taskid,
            'S_Status': self.status,
            'S_CreatedAt': self.created_at,
            'S_FinishedAt': self.finished_at,
            'N_Progress': self.progress,
            'S_Error': self.error,
        }

    @classmethod
    def from_item(cls, item: dict):
        """ Create a task item instance from dict """
        task = cls()

        task.itemid = item['S_ItemId']
        task.source = item['S_Source']
        task.taskid = item['S_TaskId']
        task.status = item.get('S_Status', None)
        task.progress = item.get('N_Progress', 0)
        task.created_at = item.get('S_CreatedAt', None)
        task.finished_at = item.get('S_FinishedAt', None)
        task.error = item.get('S_Error', None)

        return task


def get_task(taskid: str) -> Task:
    """
    Get task by taskid
    Args:
        taskid: The id of Task
    Returns:
        The task object, if id not exists in DynamoDB then return `None`
    """
    resp = db.Table(taskitem_table_name).get_item(Key={'S_TaskId': taskid})
    task = Task.from_item(resp['Item']) if resp.get('Item', None) is not None else None

    return task


def create_task(taskid: str, bucket: str, key: str = None, condition: str = None, template: str = None) -> Task:
    """
    Create task
    Args:
        taskid: The id of Task
        bucket: Bucket name where the source in
        key: Key of the source in bucket
        condition: The filter condition used to look up objects in bucket
        template: The template name used to create MediaConvert job
    Returns:
        A task object
    """
    task = Task()
    task.taskId = taskid
    task.bucket = bucket
    task.key = key
    task.filter = condition
    task.template_name = template

    db.Table(task_table_name).put_item(
        Item=task.as_dict(),
        ReturnValues='NONE'
    )

    return task


def get_task_item(itemid: str) -> TaskItem:
    """
    Get task by itemid
    Args:
        itemid: The id of Task item
    Returns:
        The task item object, if id not exists in DynamoDB then return `None`
    """
    resp = db.Table(taskitem_table_name).get_item(Key={'S_ItemId': itemid})
    item = TaskItem.from_item(resp['Item']) if resp.get('Item', None) is not None else None

    return item


def set_task_total(taskid: str, total: int):
    """
    Set total number of the task has
    Args:
        taskid: The id of Task
        total: The total number
    """
    db.Table(task_table_name).update_item(
        Key={'S_TaskId': taskid},
        UpdateExpression='SET N_Total = :total, S_ExecutedAt = :at',
        ExpressionAttributeValues={
            ':total': total,
            ':at': datetime.now().astimezone().strftime('%Y-%m-%d %H:%M:%S%z')
        },
        ReturnValues='NONE'
    )


def increase_task_running_counter(taskid: str):
    """
    Increase the running job counter of the Task
    Args:
        taskid: The id of Task
    """
    db.Table(task_table_name).update_item(
        Key={'S_TaskId': taskid},
        UpdateExpression='SET N_Running = N_Running + :one',
        ExpressionAttributeValues={':one': 1},
        ReturnValues='NONE'
    )


def increase_task_finished_counter(taskid: str):
    """
    Increase the finished job counter of the Task
    Args:
        taskid: The id of Task
    """
    db.Table(task_table_name).update_item(
        Key={'S_TaskId': taskid},
        UpdateExpression='SET N_Finished = N_Finished + :one, N_Running = N_Running - :one',
        ExpressionAttributeValues={':one': 1},
        ReturnValues='NONE'
    )


def increase_task_error_counter(taskid: str):
    """
    Increase the error job counter of the Task
    Args:
        taskid: The id of Task
    """
    db.Table(task_table_name).update_item(
        Key={'S_TaskId': taskid},
        UpdateExpression='SET N_Error = N_Error + :one, N_Running = N_Running - :one',
        ExpressionAttributeValues={':one': 1},
        ReturnValues='NONE'
    )


def update_taskitem_status(itemid: str, status: str, error: str = None):
    """
    Update the task item status
    Args:
        itemid: The id of Task item
        status: The status to update to
        error: The error infomation if has
    """
    db.Table(taskitem_table_name).update_item(
        Key={'S_ItemId': itemid},
        UpdateExpression='SET S_Status = :status, S_FinishedAt = :at, S_Error = :error',
        ExpressionAttributeValues={
            ':status': status,
            ':at': datetime.now().astimezone().strftime('%Y-%m-%d %H:%M:%S%z'),
            ':error': error,
        },
        ReturnValues='NONE'
    )


def update_taskitem_progress(itemid: str, progress: int):
    """
    Update the task item progress
    Args:
        itemid: The id of Task item
        progress: The progress of job with MediaConvert
    """
    db.Table(taskitem_table_name).update_item(
        Key={'S_ItemId': itemid},
        UpdateExpression='SET N_Progress = :progress',
        ExpressionAttributeValues={':progress': progress},
        ReturnValues='NONE'
    )


def is_taskitem_exists(bucket: str, key: str) -> bool:
    """
    Check whether the task item is exists
    Args:
        bucket: Bucket name where the source in
        key: Key of the source in bucket
    Returns:
        Whether the task item is exists
    """
    item = db.Table(taskitem_table_name).query(
        KeyConditionExpression='S_Source = :source',
        ExpressionAttributeValues={':source': get_source(bucket, key)},
        IndexName='SourceIndex',
        Select='COUNT'
    )

    return item.get('Count', 0) > 0


def is_task_exists(bucket: str, key: str, condition: str) -> bool:
    """
    Check whether the task is exists
    Args:
        bucket: Bucket name where the source in
        key: Key of the source in bucket
        condition: The filter condition used to look up objects in bucket
    Returns:
        Whether the task is exists
    """
    item = db.Table(task_table_name).query(
        KeyConditionExpression='S_Bucket = :bucket',
        IndexName='BucketIndex',
        FilterExpression='S_Key = :key AND S_Filter = :filter',
        ExpressionAttributeValues={
            ':bucket': bucket,
            ':key': key,
            ':filter': condition
        },
        Select='COUNT'
    )

    return item.get('Count', 0) > 0


def create_converter_job(taskid: str, bucket: str, key: str, template_name: str):
    """
    Create MediaConvert job, save info to taskitem and update task running/error counter
    Args:
        taskid: The id of Task
        bucket: Bucket name where the source in
        key: Key of the source in bucket
        template_name: The template name used to create MediaConvert job
    """

    increase_task_running_counter(taskid)
    source = get_source(bucket, key)
    dest = None
    error = None

    # noinspection PyBroadException
    try:
        with open('./task_params.json', 'r') as f:
            params = json.load(f)
            params['Role'] = _get_options('MediaConvertJobRole')
            params['JobTemplate'] = template_name
            params['Settings']['Inputs'][0]['FileInput'] = source

        template_params = converter.get_job_template(Name=template_name)
        dest = template_params['JobTemplate']['Settings']['OutputGroups'][0]['OutputGroupSettings'].get(
            'Destination', None
        )

        if dest is None:
            dest = _get_options('%s-OutputBucket' % bucket)
            if dest is None:
                dest = _get_options('default-OutputBucket')

            sub = ''
            if '/' in key:
                sub = key[0:key.rindex('/') + 1]

            params['Settings']['OutputGroups'] = [{
                'OutputGroupSettings': {
                    'Type': 'FILE_GROUP_SETTINGS',
                    'FileGroupSettings': {
                        'Destination': 's3://%s/%s' % (dest, sub)
                    },
                }
            }]

        resp = converter.create_job(**params)

        status = 'RUNNING'
        itemid = resp['Job']['Id']
        created_at = resp['Job']['CreatedAt'].strftime('%Y-%m-%d %H:%M:%S%z')
        finished_at = None
    except Exception as err:
        error = str(err)
        increase_task_error_counter(taskid)

        status = 'ERROR'
        itemid = uuid.uuid4().hex
        created_at = datetime.now().astimezone().strftime('%Y-%m-%d %H:%M:%S%z')
        finished_at = datetime.now().astimezone().strftime('%Y-%m-%d %H:%M:%S%z')

    db.Table(taskitem_table_name).put_item(
        Item={
            'S_ItemId': itemid,
            'S_Source': source,
            'S_Target': get_source(dest, key),
            'S_TaskId': taskid,
            'S_Status': status,
            'N_Progress': 0,
            'S_CreatedAt': created_at,
            'S_FinishedAt': finished_at,
            'S_Error': error,
        },
        ReturnValues='NONE')


def source_file_exists(bucket: str, key: str) -> bool:
    """
    Check whether the source object is exists
    Args:
        bucket: Bucket name where the source in
        key: Key of the source in bucket
    Returns:
        Whether the source object is exists
    """
    if bucket is None or len(bucket.strip()) == 0:
        return False
    if key is None or len(key.strip()) == 0:
        return False

    # noinspection PyBroadException
    try:
        f = boto3.resource('s3').Object(bucket, key)
        if f.content_length == 0:
            return False
    except:
        return False

    return True


def get_bucket_template_name(bucket: str) -> str:
    """
    Get the JobTemplate name of MediaConvert for the bucket from options
    If the bucket not has a pre-defined options, a default option will used
    Args:
        bucket: Bucket name where the source in
    Returns:
        The JobTemplate name of MediaConvert
    """
    template = _get_options('%s-JobTemplate' % bucket)
    if template is None:
        template = _get_options('default-JobTemplate')

    return template


def get_source(bucket: str, key: str) -> str:
    """
    Get a source object url
    Args:
        bucket: Bucket name where the source in
        key: Key of the source in bucket
    Returns:
        A s3 protocol url string
    """
    return "s3://%s/%s" % (bucket, key)


_options = dict()  # global option store


def _get_options(key: str) -> any:
    opt = _options.get(key, None)
    if opt is not None:
        return opt

    item = db.Table(options_table_name).get_item(
        Key={'S_Key': key}
    )

    value = item.get('Item', dict()).get('S_Value', None)
    _options[key] = value
    return value


video_converter/task_event.py
# -*- coding: utf-8 -*-

import logging
import math
from task import *

logging.getLogger().setLevel(logging.INFO)
logger = logging.getLogger(__name__)


def lambda_handler(event, _):
    logger.info("Received event: " + json.dumps(event, indent=2))

    code = 200
    error = None
    progress = None

    itemid = event['detail']['jobId']
    status = event['detail']['status']

    taskitem = get_task_item(itemid)
    if taskitem is not None:
        if status == 'COMPLETE':
            increase_task_finished_counter(taskitem.taskid)
        elif status == 'ERROR':
            error = event['detail']['errorMessage']
            increase_task_error_counter(taskitem.itemid)
        elif status == 'STATUS_UPDATE':
            progress = math.floor(float(event['detail']['jobProgress']['jobPercentComplete']))
            update_taskitem_progress(itemid, progress)

        if status != 'STATUS_UPDATE':
            update_taskitem_status(itemid, status, error)
            update_taskitem_progress(itemid, 100 if status == 'COMPLETE' else '-1')

        logger.info("Job(%s) status is updated to [%s] "
                    % (itemid, str(progress) if status == 'STATUS_UPDATE' else status))
    else:
        code = 400
        error = "Job not exists"

        logger.info("Job(%s) not exists. " % itemid)

    return {"status": code, "event": event, 'message': error}


video_converter/auto_executor.py
# -*- coding: utf-8 -*-

import urllib.parse
import logging
import traceback
from task import *


logging.getLogger().setLevel(logging.INFO)
logger = logging.getLogger(__name__)


def lambda_handler(event, _):
    logger.info("Received event: " + json.dumps(event, indent=2))

    try:
        bucket = event['Records'][0]['s3']['bucket']['name']
        key = urllib.parse.unquote_plus(event['Records'][0]['s3']['object']['key'], encoding='utf-8')

        if not key.endswith('/') and source_file_exists(bucket, key):
            # always create task when invoked by s3 event notification
            template = get_bucket_template_name(bucket)
            task = create_task(uuid.uuid4().hex, bucket, key, None, template)
            create_converter_job(task.taskId, task.bucket, task.key, task.template_name)
            set_task_total(task.taskId, 1)

            logger.info('Task created.')
            return {"status": 200, "event": event, 'message': None}
    except Exception as err:
        error = ''.join(traceback.format_exception(None, err, err.__traceback__))

        logger.error("Manual task execute error: " + error)
        return {'status': 400, 'event': event, 'message': error}

    return {"status": 400, "event": event, 'message': 'Ignored. Key is directory or task already exists.'}


video_converter/manual_executor.py
# -*- coding: utf-8 -*-

import logging
import traceback
from task import *
# from mediainfo import get_media_info

logging.getLogger().setLevel(logging.INFO)
logger = logging.getLogger(__name__)


def lambda_handler(event, _):
    logger.info("Manual task request: " + json.dumps(event, indent=2))

    msg = event['Records'][0]
    attributes = msg['messageAttributes']

    try:
        taskid = msg['messageId']

        bucket = attributes.get('Bucket', dict()).get('stringValue', None)
        key = attributes.get('Key', dict()).get('stringValue', None)
        template = attributes.get('TemplateName', dict()).get('stringValue', None)
        condition = attributes.get('Filter', dict()).get('stringValue', None)
        force = attributes.get('Force', dict()).get('stringValue', 'False').upper() == 'TRUE'

        if bucket is None:
            raise ValueError('bucket is none')

        if is_task_exists(bucket, key, condition) and not force:
            logger.info('Task already exists, exit!')
            return {'status': 400, 'event': event, 'message': 'task already exists'}

        if template is None:
            template = get_bucket_template_name(bucket)

        task = create_task(taskid, bucket, key, condition, template)
        total = 0

        if not task.key or task.key.endswith('/'):
            client = boto3.client('s3').get_paginator('list_objects_v2')
            if not task.key:
                paginator = client.paginate(Bucket=task.bucket)
            else:
                paginator = client.paginate(Bucket=task.bucket, Prefix=task.key)
            files = paginator.search(condition if condition is not None else 'Contents[]')

            for f in files:
                key = f['Key']
                if key.endswith('/'):
                    continue

                # get video resolution by mediainfo
                # mi = get_media_info(bucket, key)
                # for track in mi.tracks:
                #     if track.track_type == 'Video':
                #         logger.info('Video resolution is %d x %d' % (track.width, track.high))

                logger.info('Job recieved, source - %s' % get_source(bucket, key))

                if not key.endswith('/'):
                    if force or not is_taskitem_exists(task.bucket, key):
                        create_converter_job(task.taskId, task.bucket, key, task.template_name)
                        total += 1
                    else:
                        logger.info('Job already exists, source - %s' % get_source(bucket, key))
                else:
                    logger.info('Job source is directory, source - %s' % get_source(bucket, key))

        elif not task.key.endswith('/'):
            logger.info('Job recieved, source - %s' % get_source(bucket, key))

            if source_file_exists(task.bucket, task.key) > 0:
                if force or not is_taskitem_exists(task.bucket, key):
                    create_converter_job(task.taskId, task.bucket, task.key, task.template_name)
                    total += 1
                else:
                    logger.info('Job already exists, source - %s' % get_source(bucket, key))
            else:
                logger.info('Job source not exists, source - %s' % get_source(bucket, key))

        set_task_total(task.taskId, total)
        logger.info('Manual task started, total job - %d' % total)
    except Exception as err:
        error = ''.join(traceback.format_exception(None, err, err.__traceback__))

        # TODO log error to dynamodb

        logger.error("Manual task executed with error - " + error)
        return {'status': 400, 'event': event, 'message': error}

    return {"status": 200, "event": event, 'message': None}
